### PR TITLE
Add RenderWithFile support

### PR DESCRIPTION
**Pull request by DsXack**
Fix for #55

### CODE REVIEW

**Comment by dave** [comment]
OK this looks good, and it's a valid addition. A few things:

1) Don't need `GoStringWithFile`... `GoString` is called with the `fmt` package formats with the `#v` verb, so adding a variation doesn't make sense.

2) We should tweak the internals of the `Render` function to use `RenderWithFile(writer, f)` instead of duplicating all the logic.

3) The `Group` type has a `Render` method too... it should probably have `RenderWithFile` also.

Thanks for the PR!

**Comment by DsXack** [comment]
Thank you for review. I'm going to fix remarks now

**Comment by DsXack** [comment]
I've done your PR comments

File: jen/group.go [hunk 1]
```diff
@@ -116,8 +116,13 @@ func (g *Group) renderItems(f *File, w io.Writer) (isNull bool, err error) {
 }
 
 // Render renders the Group to the provided writer.
-func (g *Group) Render(writer io.Writer) error {
-	f := NewFile("")
+func (g *Group) Render(writer io.Writer, file ...*File) error {
```
Comment (dave) [inline jen/group.go]: It's generally though of as bad practise to use variadic parameters where there's an optional parameter, so `Render` should only the writer.

File: jen/statement.go [hunk 2]
```diff
@@ -70,8 +70,14 @@ func (s *Statement) render(f *File, w io.Writer, _ *Statement) error {
 }
 
 // Render renders the Statement to the provided writer.
-func (s *Statement) Render(writer io.Writer) error {
-	f := NewFile("")
+func (s *Statement) Render(writer io.Writer, file ...*File) error {
```
Comment (dave) [inline jen/statement.go]: ... as above.

File: jen/group_test.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+	"testing"
+
+	. "github.com/dave/jennifer/jen"
+)
+
+func TestGroup_Render(t *testing.T) {
+	file := NewFile("main")
+	file.ImportAlias("fmt", "fmtalias")
+
+	g := &Group{
```
Comment (dave) [inline jen/group_test.go]: To get a `Group` without having to use private fields, you could do something like:

```go
var group *Group
b := BlockFunc(g *Group) {
    group = g
}
group.RenderWithFile(...)
```

... I think that would work?

Comment (DsXack) [reply jen/group_test.go]: Fixed

File: jen/statement.go (hunk 2 above)
Comment (DsXack) [reply jen/statement.go]: Fixed, thanks

File: jen/group.go (hunk 1 above)
Comment (DsXack) [reply jen/group.go]: Fixed, Thanks

File: jen/statement.go [hunk 4]
```diff
@@ -69,9 +69,7 @@ func (s *Statement) render(f *File, w io.Writer, _ *Statement) error {
 	return nil
 }
 
-// Render renders the Statement to the provided writer.
-func (s *Statement) Render(writer io.Writer) error {
-	f := NewFile("")
+func (s *Statement) renderWithFile(w io.Writer, f *File) error {
```
[the remark this answers is not in the document]
Comment (DsXack) [reply jen/statement.go]: Thanks, sorry... head bad at the end of the day

[the remark this answers is not in the document]
Comment (DsXack) [reply jen/statement.go]: Fixed

File: jen/group.go [hunk 5]
```diff
@@ -118,8 +118,22 @@ func (g *Group) renderItems(f *File, w io.Writer) (isNull bool, err error) {
...
+// GoString renders the Group for testing. Any error will cause a panic.
+func (g *Group) GoString() string {
+	buf := bytes.Buffer{}
+	if err := g.Render(&buf); err != nil {
+		panic(err)
+	}
+	return buf.String()
+}
+
+// RenderWithFile renders the Statement to the provided writer
```
Comment (dave) [inline jen/group.go]: `Statement` should be `Group` in this comment. 

Also, let's change the comment to "RenderWithFile renders the Group to the provided writer, using imports from the provided file."

File: jen/statement.go [hunk 6]
```diff
@@ -72,8 +72,22 @@ func (s *Statement) render(f *File, w io.Writer, _ *Statement) error {
...
+// GoString renders the Statement for testing. Any error will cause a panic.
+func (s *Statement) GoString() string {
+	buf := bytes.Buffer{}
+	if err := s.Render(&buf); err != nil {
+		panic(err)
+	}
+	return buf.String()
+}
+
+// RenderWithFile renders the Statement to the provided writer
```
Comment (dave) [inline jen/statement.go]: Also, let's change the comment to "RenderWithFile renders the Statement to the provided writer, using imports from the provided file."

File: jen/group.go (hunk 5 above)
Comment (DsXack) [reply jen/group.go]: Fixed, thanks

File: jen/statement.go (hunk 6 above)
Comment (DsXack) [reply jen/statement.go]: Fixed, thanks

File: jen/group.go [hunk 7]
```diff
@@ -118,8 +118,22 @@ func (g *Group) renderItems(f *File, w io.Writer) (isNull bool, err error) {
 // Render renders the Group to the provided writer.
 func (g *Group) Render(writer io.Writer) error {
 	f := NewFile("")
+	return g.RenderWithFile(writer, f)
```
Comment (dave) [inline jen/group.go]: Let's combine this with the previous line:

```go
return g.RenderWithFile(writer, NewFile(""))
```

File: jen/statement.go [hunk 8]
```diff
@@ -72,8 +72,22 @@ func (s *Statement) render(f *File, w io.Writer, _ *Statement) error {
 // Render renders the Statement to the provided writer.
 func (s *Statement) Render(writer io.Writer) error {
 	f := NewFile("")
+	return s.RenderWithFile(writer, f)
```
Comment (dave) [inline jen/statement.go]: As above.

File: jen/group.go (hunk 1 above)
Comment (DsXack) [reply jen/group.go]: Yea, fixed

File: jen/statement.go (hunk 2 above)
Comment (DsXack) [reply jen/statement.go]: Fixed

File: jen/group.go (hunk 7 above)
Comment (DsXack) [reply jen/group.go]: Fixed

File: jen/statement.go (hunk 8 above)
Comment (DsXack) [reply jen/statement.go]: Fixed

**Comment by dave** [comment]
Merged! Thanks!

**Comment by DsXack** [comment]
Thank you!

**Comment by dave** [comment]
I've tagged this release as `v1.3.0`.

Thanks very much!

**Pull request merged**